### PR TITLE
Name (un-renamed) file on first save

### PR DESCRIPTION
**Pull request by jess-x**
## References

per #5207, to mitigate having a bunch of Untitled files,
this PR makes a Name File popup to rename (Untitled) files on first save and a toggle in settings

## User-facing changes
ezgif-2-6e792d39b4cf

in progress
- [x] fix dirty flag
- [x] fix bug in active document
- [x] checkbox styling
- [x] fix bug in toggle 
- [x] generalize document type
- [x] update and add comments
- [x] settings validate additional properties
- [x] Auto save

### CODE REVIEW

**Comment by goanpeca** [comment]
Great work @jess-x 

It seems after I hit the first save, I get the dialog (yeah!) but after saving, I still see the notebook as "dirty" (unsaved/midifed state)

See the animation below:

test

**Comment by goanpeca** [comment]
Should we have the text next to the checkbox @jess-x ?

What do you think @isabela-pf  ?

image

**Comment by goanpeca** [comment]
Also if I open a new notebook and close the file (by clicking the x on the tab) I get the dialog that says, save.discard.

I was expecting that when I hit save, it would prompt me with the dialog to put the name

Thoughts @isabela-pf ?

File: packages/docmanager-extension/src/index.ts
Comment (goanpeca) [inline packages/docmanager-extension/src/index.ts:188]: Not sure I understand this logic here @jess-x could you explain?

**Comment by jess-x** [comment]
> It seems after I hit the first save, I get the dialog (yeah!) but after saving, I still see the notebook as "dirty" (unsaved/midifed state)
Thanks @goanpeca,
I couldn't reproduce this, for me it saves correctly. The dialog shows the first save, not the second, but on the third save. 
gif:

**Comment by goanpeca** [comment]
@jess-x if I check the checkbox, what is the expected result? Should this affect the settings directly?

Change the value from `true` to `false`?

test3

File: packages/docmanager/src/dialogs.ts
Comment (goanpeca) [inline packages/docmanager/src/dialogs.ts:309]: Should we add the label next to the checkbox @jess-x ?

File: packages/docmanager/src/manager.ts
Comment (goanpeca) [inline packages/docmanager/src/manager.ts:136]: Is this stored per document or per global setting. It feels like it should be a setting?

File: packages/docmanager/src/dialogs.ts
Comment (jess-x) [reply packages/docmanager/src/dialogs.ts:309]: We should, and I was researching a way to do that, any suggestions @goanpeca ?

Comment (goanpeca) [reply packages/docmanager/src/dialogs.ts:309]: Create a div and add to the div, and then add the div as a child?

Comment (goanpeca) [reply packages/docmanager/src/dialogs.ts:309]:
```ts
    const trans = translator.load('jupyterlab');
    const body = document.createElement('div');
    const name = document.createElement('input');
    const checkbox = document.createElement('input');
    const div = document.createElement('div');
    const label = document.createElement('label');

    checkbox.type = 'checkbox';
    checkbox.addEventListener('change', function () {
      manger.nameFileOnSave = !this.checked;
    });
    // Question: aglin two elements (flexbox?)
    label.textContent = trans.__("Don't ask me again");

    div.appendChild(checkbox);
    div.appendChild(label);

    body.appendChild(name);
    body.appendChild(div)
```

File: packages/docmanager/src/dialogs.ts
Comment (goanpeca) [inline packages/docmanager/src/dialogs.ts:291]: manager

File: packages/docmanager/src/manager.ts
Comment (jess-x) [reply packages/docmanager/src/manager.ts:136]: @goanpeca you mentioned Signal to settingRegister on change in nameFileOnSave. I am actually not too sure how that work. any pointer to an example maybe?

Comment (goanpeca) [reply packages/docmanager/src/manager.ts:136]: See this line, https://github.com/jupyterlab/jupyterlab/pull/10043/files#diff-e0f9ab316153ea5eac9b4e2757412d93bbd75b744c89e8c88818f7ce795bd2d7L42

We could add a new signal to the interface. And make make it more generic like `optionChanged` ?

`readonly optionChanged: ISignal<this, Object>;`

then on the set method

```ts
set nameFileOnSave(value: boolean) {
    this._nameFileOnSave = value;
    this.optionChanged.emit({nameFileOnSave: value});
}
```


And then on the docmanager-extension file we could do something like:


```ts
    docManager.optionChanged.connect(...)
```

Does this make sense?

Could you also look at how other components implement `IOptions` that way we can follow the same paradigm for the constructor?

**Comment by isabela-pf** [comment]
Thanks for the mention! Here are my thoughts.

> Should we have the text next to the checkbox @jess-x ?

This was already answered, but I agree we should.

> Also should the text be "Save"?

Yup! That is more clear and it matches the patterns for button text elsewhere in JupyterLab.

> Should there be a "Cancel" button?

There definitely needs to be some way to escape the modal dialog, and "Cancel" is what we are using elsewhere. I think the `Save as` dialog is a close comparison.

Screen Shot 2021-04-06 at 3 50 57 PM

However, I think there is a little confusion with what a "Cancel" button would do with the dialog currently being named "Name file." As far as I can tell, this PR makes it so that saving a file automatically brings users to both rename and save a file (just like `Save As`). I think seeing "Cancel" with that title would make me wonder if I was just cancelling the rename, cancelling the save, or both.

I'd prefer making the title of the dialog to be "Save File As" instead of "Name file" because that seems both more accurate for the behavior and consistent with an experience users already know. Then users should be more clear on what they are cancelling. Let me know if I'm misunderstanding anything.

> Should this affect the settings directly?

I would expect it to, but that's just from my perspective. I'd be open to hearing if there's a reason the settings shouldn't be updated accordingly.

I have one more not not related to any feedback already given. Is there any way this can be done so *only* first save of a new file prompts a name-and-save experience? Even without users selecting the checkbox? It seems overdone and disruptive to ask a user to suddenly have a `Save as` interaction for every single save even if they don't configure it.

**Comment by jess-x** [comment]
@isabela-pf  
> Is there any way this can be done so only first save of a new file prompts a name-and-save experience? Even without users selecting the checkbox?
Yes, I think that's the least intrusive to the current user experience and and most intuitive. In fact, this PR is intended to have the popup only appear on the first save, however at this time is still work in progress unfortunately.

For change Name File to Save As, I can see a make a lot more sense for the switch. I thought since the Save and Name file is two separate processes, a failure in Name File (either with hitting the cancel button or entering an invalid file name) may convey to the user there is a failure in Save, but I think this can easily mitigated by a custom error message e.g."named file failed, saved successfully" this way can convey separate mental models for the two processes?

File: packages/docmanager/src/manager.ts
Comment (jess-x) [reply packages/docmanager/src/manager.ts:136]: @goanpeca thx for the info! I tired it out and looks like using Signal has a lot of potentials. and Yes thanks for suggesting `IOptions`.

**Comment by isabela-pf** [comment]
> at this time is still work in progress unfortunately.

Not unfortunate, it's my misunderstanding. Thanks for clarifying!

> I thought since the Save and Name file is two separate processes

If they really are separate in this case then I agree they should have different names. I could be misunderstanding again, but it looked like saving triggered the naming process and there was no additional feedback for the save so they (from the user side) look like one process. 

So it is possible with this work for a rename to fail but the save to succeed, or the opposite? In general I'd prefer not to have an additional error message because in JupyterLab that would mean putting two dialogs back-to-back. Do you know what the current feedback is for a failed save? I haven't been able to create that situation to see what it does.

**Comment by jess-x** [comment]
@isabela-pf thanks so much for your thoughts into this! 

As for now, I have moved the Name File popup to whenever a file is *successfully* saved for the first time, so hopefully this mitigates the annoyance of having to deal with all the edge cases where one process successes while the other fails. 

However, the question remains: What is best interaction between Name File and the current Discard Unsaved Changes popup. I see there are two potential solutions to this:
1.  Integrate Name File into the current Discard Unsaved Changes popup
2. Skip the Name File process when you close a file with unsaved changes, in favor of the *current*  Discard Unsaved Changes popup (I *think* most people will scape the unsaved changes anyways)

For the first option, the best approach I could think of:
- In the current Discard Unsaved Changes popup, add a Name File input box but is greyed out by default, and is activated when "save" is hit 

However, even this I think gives folks too many decisions to make when simply just trying close a file. So I am actually leaning more toward the second option.
Thoughts on this @isabela-pf?

**Comment by isabela-pf** [comment]
I know I initially said I thought combining them was best (your first option), but I think I agree skipping (the second option) might be simpler. I don't know how often people intend to scrap changes versus just forget to save, but I do think saving or discarding changes is a much more important choice for the user to be making in that case.

To help me decide, I made a quick mockup of what I think a combined dialog might look like. I totally agree there's a lot (maybe too much) going on here. I've attached it here in case it helps anyone else.
Screen Shot 2021-04-29 at 4 19 26 PM

File: packages/notebook/src/panel.ts
Comment (goanpeca) [inline packages/notebook/src/panel.ts:110]: Just wondering if this will be robust enough, or wether we should take this from the place it is defined? (it is an option that is configurable, right?)

The `Untitled` I mean

**Comment by goanpeca** [comment]
Awesome work @jess-x. Could you fix the merge conflicts :) 

Thanks!

File: packages/notebook/src/default-toolbar.tsx
Comment (goanpeca) [inline packages/notebook/src/default-toolbar.tsx:72]: Was this space planned :) ?

**Comment by goanpeca** [comment]
Works like a charm @jess-x 

Minor nitpick, could we align the checkbox and the text vertically? And maybe also add a space between them?


image


Thanks!

File: packages/notebook/src/panel.ts
Comment (jess-x) [reply packages/notebook/src/panel.ts:110]: it might not be robust code-wise, but I thought, because the Name File popup is to avoid having the `Untitled`s (and I am assuming if folks are changing `Untitled` to something else, they might want to keep it, instead of renaming). So I am hoping using `Untitled` would be sufficient for now?

File: packages/notebook/src/default-toolbar.tsx
Comment (jess-x) [reply packages/notebook/src/default-toolbar.tsx:72]: Thx! good catch

File: packages/notebook/src/panel.ts
Comment (goanpeca) [reply packages/notebook/src/panel.ts:110]: Yeah it should be good for now

**Comment by jess-x** [comment]
@goanpeca I added some finishing touches, and the popup now looks like this:
Screen Shot 2021-05-03 at 3 31 52 PM

**Review by goanpeca — APPROVED** [review]
Looks awesome :) 🚀 

Ready for merge on my side :)

**Comment by blink1073** [comment]
Looks like there is one more merge conflict on this one.

**Comment by goanpeca** [comment]
>Looks like there is one more merge conflict on this one.

Fixed!

**Comment by blink1073** [comment]
"Please run `jlpm run integrity` locally and commit the changes"

File: packages/docmanager-extension/src/index.ts
Comment (goanpeca) [inline packages/docmanager-extension/src/index.ts:781]: @jess-x shouldn't we be doing this for all files, not just notebooks? 

I am also wary of adding this dependency to docmanager:

**Comment by isabela-pf** [comment]
Just following up, what's still blocking this PR?

I was in a meeting where people were talking about the number of "untitled" notebooks they had and it reminded me to check in on this lovely work.

**Comment by goanpeca** [comment]
Pinging @jupyterlab/committers :)

**Comment by krassowski** [comment]
It is a wonderful work and the from quick skimming the code looks good to me (I have not analysed it in detail though). There are two potential improvements I could think of:
1. smart handling of auto-save
2. future-proofing the API by adding some way to pass extra context data to `shouldNameFile` signal

(1) Currently the dialog will show up on auto-save too. I would prefer that it only shows up on manual save. To reproduce modify settings in "Document manager" tab of Advanced Settings Editor to:

```json
{
    "autosaveInterval": 1,
}
```

Then:
- create and open a new notebook
- wait for 3 seconds without doing anything
- see the rename dialog show up (unexpectedly)

Some users might be annoyed by this behaviour, even if they like the idea of having the prompt on manual save. Should we modify the save signal to include information on whether it was manual or not and only display the dialog on manual save?

(2) It might be good to hook a similar behaviour with the "Share Link" function for collaborative editing; I like that `shouldNameFile` signal is public - this should allow to do so; to make it better, it could accept some extra context information e.g a string describing why the dialog is being showed up, this is:

`shouldNameFile: Signal<any, string>`

and it would be emitted like:

- `shouldNameFile.emit('Name File on Save')` - in this PR
- `shouldNameFile.emit('Name File Before Sharing')` - hypothetically in the share link extension

This is just an idea on making the API future-compatible, I would wait for others to comment on whether it is a good one before implementing it ;)

**Review by blink1073 — APPROVED** [review]
Fantastic, thank you @jess-x!

File: packages/notebook/src/panel.ts
Comment (jtpio) [reply packages/notebook/src/panel.ts:111]: Looks like this is configurable with `ContentsManager.untitled_file`:

https://jupyter-server.readthedocs.io/en/latest/full-config.html?#options

**Pull request merged**
